### PR TITLE
PHP 8.0 and PHP 8.1 support, BetterReflection 5.x upgrade

### LINKED ISSUE #283: PHP 8 support
Aside from #264, is there anything else that is blocking a PHP 8 compatible release? 

P.S. This is a tracking issue for a release that supports PHP 8.

### PR DESCRIPTION
**Pull request by Ocramius**
This patch brings in PHP 8.0 and 8.1 support through the massive work done by @kukulich on
`roave/better-reflection` over the past few months.

The upgrade is mostly smooth, but we have some rough edges with:

 * [x] `TypeIsCovariant`, which needs `ReflectionType`s as well as `Reflector`s
 * [x] `TypeIsContravariant`, which needs `ReflectionType`s as well as `Reflector`s
 * [x] `ReturnTypeCovarianceChanged` uses hacky reflection to get a `Reflector` from a `ReflectionFunctionAbstract` (internal state!) - needs https://github.com/Roave/BetterReflection/pull/836
 * [x] `ParameterTypeContravarianceChanged` uses hacky reflection to get a `Reflector` from a `ReflectionFunctionAbstract` (internal state!) - needs https://github.com/Roave/BetterReflection/pull/836
 * [x] we miss tests for covariance/contravariance with `ReflectionUnionType` and `ReflectionIntersectionType` (`infection/infection` will highlight that)
 * [x] https://github.com/Roave/BetterReflection/pull/902
 * [x] tests for covariance/contravariance of nullable `ReflectionIntersectionType` (needs https://github.com/Roave/BetterReflection/pull/902 )

Fixes #283

### CODE REVIEW

File: composer.json
Comment (Ocramius) [inline composer.json:11]: We need a stable release of `roave/better-reflection` before merging this

File: composer.json
Comment (Ocramius) [inline composer.json:54]: Unsure if needed: will prevent invalid dependency upgrades though

File: src/CompareApi.php
Comment (Ocramius) [inline src/CompareApi.php:19]: Loads of inherited BC breaks because of this, but they are very much acceptable

File: src/DetectChanges/BCBreak/FunctionBased/ParameterTypeChanged.php
Comment (Ocramius) [inline src/DetectChanges/BCBreak/FunctionBased/ParameterTypeChanged.php:85]: `ReflectionType#__toString()` now includes nullability as per `roave/better-reflection:5.0.0` - this is a BC break that led to some simplification and also code changes all around the place.

File: src/DetectChanges/BCBreak/FunctionBased/ParameterTypeContravarianceChanged.php
Comment (Ocramius) [inline src/DetectChanges/BCBreak/FunctionBased/ParameterTypeContravarianceChanged.php]: May the gods have mercy of my.

This is ugly, but it highlights a design limitation when trying to traverse a whole reflection graph.

Perhaps @kukulich has a better idea on how to achieve this, but the problem is basically that we're trying to find `parent` classes and related interfaces by their name, and it is not possible to do so from a `ReflectionType` from `roave/better-reflection:5.0.0`, as `ReflectionType` is mostly immutable.

Instead, we have to fetch the associated reflector here.

Injecting the reflector upfront is not feasible either here, with the current design (the reflector is built at runtime, by some services that do run a `composer install`) :thinking:

File: src/DetectChanges/BCBreak/FunctionBased/ReturnTypeCovarianceChanged.php
Comment (Ocramius) [inline src/DetectChanges/BCBreak/FunctionBased/ReturnTypeCovarianceChanged.php]: Similar to above: this is currently the "cleanest" approach, but it highlights the limitation of the current `ReflectionType` design.

File: src/DetectChanges/Variance/TypeIsContravariant.php
Comment (Ocramius) [inline src/DetectChanges/Variance/TypeIsContravariant.php]: This will be picked up by `infection/infection`, but basically, I did **not** test beyond `ReflectionNamedType` for now.

File: src/DetectChanges/Variance/TypeIsContravariant.php
Comment (Ocramius) [inline src/DetectChanges/Variance/TypeIsContravariant.php]: TBD

File: src/DetectChanges/Variance/TypeIsCovariant.php
Comment (Ocramius) [inline src/DetectChanges/Variance/TypeIsCovariant.php]: TBD

File: src/DetectChanges/Variance/TypeIsCovariant.php
Comment (Ocramius) [inline src/DetectChanges/Variance/TypeIsCovariant.php]: This will be picked up by `infection/infection`, but basically, I did **not** test beyond `ReflectionNamedType` for now.

**Comment by Ocramius** [comment]
Test failures are caused by `Psl\Iter\Iterator::current()` not having `: mixed` declared:

```
Deprecated: Return type of Psl\Iter\Iterator::current() should either be compatible with Iterator::current(): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /github/workspace/vendor/azjezz/psl/src/Psl/Iter/Iterator.php on line 90
```

Problem is: `azjezz/psl:1.9.0` has `php:~8.0.0 || ~8.1.0` support, while the current `azjezz/psl:2.0.0` has `php:~8.1.0` support.

@azjezz do you think we could fix that in `azjezz/psl:1.9.1` perhaps?

**Comment by azjezz** [comment]
~1.9.1 is on the way 👍~

done https://github.com/azjezz/psl/releases/tag/1.9.1

File: src/DetectChanges/BCBreak/FunctionBased/ParameterTypeContravarianceChanged.php
Comment (kukulich) [reply src/DetectChanges/BCBreak/FunctionBased/ParameterTypeContravarianceChanged.php]: https://github.com/Roave/BetterReflection/pull/836

**Comment by Ocramius** [comment]
Whoa, that was fast, @azjezz and @kukulich! 

I'll work on this some more tonight, trying out both patches ASAP! 💪

**Comment by Ocramius** [comment]
Interesting, PHPUnit with lowest dependencies still fails :thinking: 

EDIT:

```
Deprecated: Return type of Symfony\Component\Process\Process::getIterator($flags = 0) should either be compatible with IteratorAggregate::getIterator(): Traversable, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /github/workspace/vendor/symfony/process/Process.php on line 612

Call Stack:
    0.0001     389552   1. {main}() /github/workspace/bin/roave-backward-compatibility-check:0
    0.0006     464800   2. require_once('/github/workspace/bin/roave-backward"
```

I'll add a `"conflict": {"symfony/process": "<5.3.7"}`

**Comment by Ocramius** [comment]
Funny, lowest dependency runs now run forever:

Selection_824

File: src/DetectChanges/Variance/TypeWithReflectorScope.php
Comment (Ocramius) [inline src/DetectChanges/Variance/TypeWithReflectorScope.php]: This entire file is no longer needed after https://github.com/Roave/BetterReflection/pull/836

**Comment by Ocramius** [comment]
@WyriHaximus since this is your favorite overall github project, can I have your review too, please? :D

**Comment by WyriHaximus** [comment]
@Ocramius Sure, it seems that your CI is failing 😜 . Just kidding, let me go over the thing :)

**Comment by Ocramius** [comment]
Haha, yeah, CI is failing because the `6.0.x` branch contains BC breaks (compared to the last `5.0.x` release), so that bit is kinda normal

File: composer.json
Comment (WyriHaximus) [reply composer.json:54]: That is exactly how I'm using it 👍

File: src/Changes.php
Comment (WyriHaximus) [inline src/Changes.php:81]: It is more of a learning moment for myself: but why did you make this change?

**Review by WyriHaximus — APPROVED** [review]
Not perse my overall favourite project but I'm using it for some cool stuff https://github.com/WyriHaximus/php-test-utilities/pull/260/files#diff-c8f1537d7020a410a1f19233721bb24de6aa93df4ef9b40795453c1647ed5f99R31 and some other stuff is blocked by it https://github.com/WyriHaximus/php-test-utilities/pull/332

**Comment by WyriHaximus** [comment]
@Ocramius Ow yeah, I didn't even look at that. Just wanted to joke about it :P. Jokes aside, this is one reason why I want to use this package more: Make it comment on PR's about what BC would be introduced where.

File: src/Changes.php
Comment (Ocramius) [reply src/Changes.php:81]: Oh, good point, this could stay iterable! 🤔

Comment (Ocramius) [reply src/Changes.php:81]: Checked: this is necessary for PHP 8.1 compatibility, where `IteratorAggregate#getIterator(): Traversable` is declared.

The change is therefore valid, and shouldn't pose a downstream BC break, since `Traversable` is covariant with `iterable` here

File: composer.json
Comment (Ocramius) [reply composer.json:11]: Will defer this to a new issue

**Comment by Ocramius** [comment]
:ship:

**Pull request merged**
